crdt/causal_tree/causal_tree.go

```diff
@@ -0,0 +1,613 @@
+	if atomID.Timestamp == 0 {
+		return -1
+	}
+	for i, atom := range t.Weave {
+		if atom.ID == atomID {
+			return i
+		}
+	}
+	return len(t.Weave)
+}
+
+// Gets an atom from yarns.
+//
+// Time complexity: O(1)
+func (t *CausalTree) getAtom(atomID atm.AtomID) atm.Atom {
+	return t.Yarns[atomID.Site][atomID.Index]
+}
+
+// Inserts an atom in the given weave index.
+//
+// Time complexity: O(atoms)
+func (t *CausalTree) insertAtom(atom atm.Atom, i int) {
+	t.Weave = append(t.Weave, atm.Atom{})
+	copy(t.Weave[i+1:], t.Weave[i:])
+	t.Weave[i] = atom
+}
+
+// +------+
+// | Fork |
+// +------+
+
+// Fork a replicated tree into an independent object.
+//
+// Time complexity: O(atoms)
+func (t *CausalTree) Fork() (*CausalTree, error) {
+	if len(t.Sitemap)-1 >= math.MaxUint16 {
+		return nil, ErrSiteLimitExceeded
+	}
+	newSiteID := uuidv1()
+	i := siteIndex(t.Sitemap, newSiteID)
+	if i == len(t.Sitemap) {
+		t.Yarns = append(t.Yarns, nil)
+		t.Sitemap = append(t.Sitemap, newSiteID)
+	} else {
+		// Remap atoms in yarns and weave.
+		localRemap := make(indexmap.IndexMap)
+		for j := i; j < len(t.Sitemap); j++ {
+			localRemap.Set(j, j+1)
+		}
+		for i, yarn := range t.Yarns {
+			for j, atom := range yarn {
+				t.Yarns[i][j] = atom.RemapSite(localRemap)
+			}
+		}
+		for i, atom := range t.Weave {
+			t.Weave[i] = atom.RemapSite(localRemap)
+		}
+		t.Cursor = t.Cursor.RemapSite(localRemap)
+		// Insert empty yarn in local position.
+		t.Yarns = append(t.Yarns, nil)
+		copy(t.Yarns[i+1:], t.Yarns[i:])
+		t.Yarns[i] = nil
+		// Insert site ID into local sitemap.
+		t.Sitemap = append(t.Sitemap, uuid.Nil)
+		copy(t.Sitemap[i+1:], t.Sitemap[i:])
+		t.Sitemap[i] = newSiteID
+	}
+	// Copy data to remote tree.
+	n := len(t.Sitemap)
+	t.Timestamp++
+	remote := &CausalTree{
+		Weave:     make([]atm.Atom, len(t.Weave)),
+		Cursor:    t.Cursor,
+		Yarns:     make([][]atm.Atom, n),
+		Sitemap:   make([]uuid.UUID, n),
+		SiteID:    newSiteID,
+		Timestamp: t.Timestamp,
+	}
+	copy(remote.Weave, t.Weave)
+	for i, yarn := range t.Yarns {
+		remote.Yarns[i] = make([]atm.Atom, len(yarn))
+		copy(remote.Yarns[i], yarn)
+	}
+	copy(remote.Sitemap, t.Sitemap)
+	return remote, nil
+}
+
+// +-------+
+// | Merge |
+// +-------+
+
+// Time complexity: O(sites)
+func mergeSitemaps(s1, s2 []uuid.UUID) []uuid.UUID {
+	var i, j int
+	s := make([]uuid.UUID, 0, len(s1)+len(s2))
+	for i < len(s1) && j < len(s2) {
+		id1, id2 := s1[i], s2[j]
+		order := bytes.Compare(id1[:], id2[:])
+		if order < 0 {
+			s = append(s, id1)
+			i++
+		} else if order > 0 {
+			s = append(s, id2)
+			j++
+		} else {
+			s = append(s, id1)
+			i++
+			j++
+		}
+	}
+	if i < len(s1) {
+		s = append(s, s1[i:]...)
+	}
+	if j < len(s2) {
+		s = append(s, s2[j:]...)
+	}
+	return s
+}
+
+// Merge updates the current state with that of another remote tree.
+// Note that merge does not move the cursor.
+//
+// Time complexity: O(atoms^2 + sites*log(sites))
+func (t *CausalTree) Merge(remote *CausalTree) {
```

Comment:
Code coverage has accused that `atom.MergeWeaves` is not covered, you should use it here.